Data/ByteString/Lazy/ReadNat.hs

```diff
@@ -0,0 +1,252 @@
+-- pairwise merging of the tail followed by a final multiplication by @10^cnt@
+-- and addition of the head.
+--
+natdigits :: ForeignPtr Word8 -- ^ Input chunk
+          -> Int              -- ^ Chunk length
+          -> Word             -- ^ accumulated element
+          -> Int              -- ^ partial digit count
+          -> [Natural]        -- ^ accumulated MSB elements
+          -> Result
+{-# INLINE natdigits #-}
+natdigits fp len = \ acc cnt ns ->
+    BI.accursedUnutterablePerformIO $
+        BI.unsafeWithForeignPtr fp $ \ ptr -> do
+            let end = ptr `plusPtr` len
+            go ptr end acc cnt ns ptr
+  where
+    go !start !end = loop
+      where
+        loop :: Word -> Int -> [Natural] -> Ptr Word8 -> IO Result
+        loop !acc !cnt ns !ptr = getDigit >>= \ !d ->
+            if | d > 9
+                 -> return $ Result (ptr `minusPtr` start) acc cnt ns
+               | cnt < safeLog
+                 -> loop (10*acc + d) (cnt+1) ns $ ptr `plusPtr` 1
+               | otherwise
+                 -> let !acc' = toNatural acc
+                     in loop d 1 (acc' : ns) $ ptr `plusPtr` 1
+          where
+            getDigit | ptr /= end = fromDigit <$> peek ptr
+                     | otherwise  = pure 10  -- End of input
+            {-# NOINLINE getDigit #-}
+            -- 'getDigit' makes it possible to implement a single success
+            -- exit point from the loop.  If instead we return 'Result'
+            -- from multiple places, when 'natdigits' is inlined we get (at
+            -- least GHC 8.10 through 9.2) for each exit path a separate
+            -- join point implementing the continuation code.  GHC ticket
+            -- <https://gitlab.haskell.org/ghc/ghc/-/issues/20739>.
+            --
+            -- The NOINLINE pragma is required to avoid inlining branches
+            -- that would restore multiple exit points.
+
+-- | Merge least-significant word with reduction of of little-endian tail.
+--
+-- The input is:
+-- * Least significant digits as a 'Word' (LSW)
+-- * The number digits that went into the LSW
+-- * All the remaining digit groups ('safeLog' digits each),
+--   in little-endian order
+--
+-- The result is obtained by pairwise recursive combining of all the
+-- full size digit groups, followed by multiplication by @10^cnt@ and
+-- addition of the LSW.
+--
+combine :: Word      -- ^ value of LSW
+        -> Int       -- ^ count of digits in LSW
+        -> [Natural] -- ^ tail elements (base @10^'safeLog'@)
+        -> Natural
+{-# INLINE combine #-}
+combine !acc !_   [] = toNatural acc
+combine !acc !cnt ns = toNatural (10^cnt) * combine1 safeBase ns + toNatural acc
+
+-- | Recursive reduction of little-endian base `base` sequence
+combine1 :: Natural -> [Natural] -> Natural
+combine1 _    [n] = n
+combine1 base ns  = combine1 (base * base) (combine2 base ns)
+
+-- | One round pairwise merge of numbers in base `base`.
```

Comment:
In haddock markup, backticks and single quotes are used for marking identifiers for hyperlinking.

```suggestion
-- | One round pairwise merge of numbers in base @base@.
```

Reply:
Note that these haddocks are internal only, they don't appear in the public API.